### ISSUE TITLE
Enhancement: Copy select links to searches & items in searches/collections

### ISSUE THREAD

**Issue by qqobb** [willsALMANJ/Zutilo#105]
This is a follow-up of #91. It would be great if Zutilo could provide <code>zotero://select</code> links for:
1. a saved search
2. items in a collection or saved search

Such links have been added to Zotero proper recently, see  https://forums.zotero.org/discussion/73893/zotero-select-for-collections. These links allow to select an item in some context, rather than in "My Library" with many unrelated items.

For 1., the links would have this form: <code>zotero://select/library/searches/[searchKey]</code>.

For 2., Zutilo could offer
(a) "Copy select item link in library/group" and
(b) "Copy select item link in collection/search"
in the context menu of an item.

Case (a) would correspond to the links currently offered by Zutilo, which now should have the form <code>zotero://select/library/items/[itemKey]</code>.

For case (b), the collectionKey or searchKey for the current view would need to be determined, and links should, e.g., look like this: 
<code>zotero://select/library/searches/[searchKey]/items/[itemKey]</code>. 

In case more than one item is selected, the links could, e.g., be 
<code>zotero://select/library/items?itemKey=[itemKey1],[itemKey2]</code>
<code>zotero://select/library/searches/[searchKey]/items?itemKey=[itemKey1],[itemKey2]</code>. 

See the <a href="https://forums.zotero.org/discussion/comment/337102/#Comment_337102">Zotero forum</a> for a list of such links. Similar links should work for items in groups, e.g.:
<code>zotero://select/groups/[groupID]/collections/[collectionKey]/items?itemKey=[itemKey1],[itemKey2]</code>. For groups, the groupID would need to be determined in addition. But this should still fall into the same scheme of (a) and (b).

**Comment by bjohas** [comment]
@willsALMANJ, @retorquere - is this something that could be solved using a translator only? This would depend on whether information about 'where the right click occurred' is passed to the translator? That may not be the case, in which case this would have to be done using a Zutilo function. (For reference: ttps://forums.zotero.org/discussion/comment/337102/#Comment_337102 )

**Comment by bjohas** [comment]
If this works via translators, as per previous comment, then this is partially solved. Either way, I have a proposal for this, related to https://github.com/willsALMANJ/Zutilo/issues/91 (copying links for collections) and https://github.com/willsALMANJ/Zutilo/issues/121 (labelling quick-copy items).

My premise is that: There are many different variations that people might want. For example, for the 'EdTech Hub, we are using kerko and want to create links like http://docs.edtechhub.org/lib/?ref=JGPWUWD5&sort=date_desc (from collections) and links like this http://docs.edtechhub.org/lib/?ref=JGPWUWD5&page=9&page-len=1&sort=date_desc for items. Because this is bespoke we cannot do this as PRs in Zutilo. 

Here's a proposal:

**Part 1. Pass selected collections/search etc to translator.** Does the translator passes the selected collections. If yes, all good. If no, then my proposal is to add this via Zutilo: I.e., a right-click passes the selected collection(s) via the Zotero item json, as an extra field. Would this be acceptable? It's kinda odd I guess as it's no longer the metadata that is being passed, but other information too. Maybe it would all be part of a sub-field called: "clickcontext" or similar? A translator could then be aware of this clickcontext, and process it if wanted. Other translators would just ignore this, so it would do no harm?

**Part 2. Allow translators for colletions/search clicks.** The solution in https://github.com/willsALMANJ/Zutilo/issues/91 'hardwires' the generation of the links, which is similar to functionality that existed in Zutlio for items. I propose to add a 'quickcopy' action (or actions) here, that gets passed the data. In the case where a collection/search is selected, this would pass a Zotero item with empty metadata, which only has a "clickcontext". This could then also be handled by a bespoke translator. (Or indeed - if you wanted to - by the same translator from Part 1, as that would either see item metadata = click on item, or it would see no item metadata = click on collection or search).

**Part 3. Do https://github.com/willsALMANJ/Zutilo/issues/121 (labelling)*** Then if the labelling is implemented, people can label the translators as they like (or indeed build setting of custom translators into their extensions (such as https://github.com/edtechhub/zotero-edtechhub).

What do you think?

**Comment by retorquere** [comment]
> **Part 1. Pass selected collections/search etc to translator.** Does the translator passes the selected collections.

It does not. You get subcollections, but the actual collection you selected and exported is not information that's available to a translators. For searches, the translator just gets  a list of items, no information about the search (or even that it was a search) is available to the translator. 

> a right-click passes the selected collection(s) via the Zotero item json, as an extra field... Maybe it would all be part of a sub-field called: "clickcontext" or similar?

I'm not sure how this would work (and I have a little knowledge of how this works). Once you click an option in the collection right-click menu, it starts a chain of events, but at the first step in that chain, the information that it was a collection, or a search, is lost. By the time you get to ItemToExportFormat (which is where you'd have to patch in the information, which is possible because I do it), I don't see any way you can trace back for what reason you got there. Nothing that's in any way robust in any case.

> In the case where a collection/search is selected, this would pass a Zotero item with empty metadata, which only has a "clickcontext".

This will likely break many export translators in a way that would be hard to debug for Zotero unless they were well aware of this behavior. Translators have a reasonable expectation that all items it can fetch are Zotero items. It certainly would break mine unless I patch them for this. You can smuggle in data by adding extra `getHiddenOption` items that wouldn't break anything, but it'd get pretty ugly pretty soon.

>  Part 2 ... add a 'quickcopy' action (or actions) here, that gets passed the data.

That still leaves another option -- new menu items added there would still have the context data, and might be able to *post-*, not pre-process the data returned from a translator. Not necessarily easy, but wouldn't break anything.

Adding a translator will add this translator as a visible option in the regular export dialog BTW. No way around that.

**Comment by bjohas** [comment]
Just to make sure I have understood. Atm, the Zotero translator is passed a structure as shown below. 
```
{
    "key": "W25GTPJU",
    "version": 4774,
    "library": {
    },
    "links": {
    },
    "meta": {
    },
    "data": {
    }
}
```
Would it really trip up other translators if another field was inserted? I.e.
```
   "clickcontext": {
   }
```
?

Let's assume we only use 'quickcopy' and assume the 'translators is triggered through quickcopy' (but not by anything else).

In other words: Zutilo offers a keyboard binding / right-click option for the standard quickcopy as well as two extra ones. (Right-click or keyboard shortcut.) If Zutilo is installed, then (only) if the right-click/keyboard shortcut is used, the additional 'clickcontext' is passed to the translator. E.g., the user may have the quickcopy set up for scannable cite. Would the extra data trip up scannable cite?

If yes, how about this: this is likely of somewhat specialist interest anyway, and also would need to be configured through the advanced settings; let's add an advanced option to decide whether the extra data is passed. For each quick-copy option, we'd need a label anyway (two options). Adding the switch, each quick-copy opton has three elements, that would look like this:
```
extensions.zutilo.quickCopy_alt1 -> export=XYZ [unchanged]
extensions.zutilo.quickCopy_alt1_label -> My export [new, #121]
extensions.zutilo.quickCopy_clickcontext-> true|false [new]
```
So, if extensions.zutilo.quickCopy_clickcontext=false, nothing else is passed and the XYZ translator has no problem. But if extensions.zutilo.quickCopy_clickcontext=true, the "clickcontext"-field is created. In that case, the XYZ translator can pick it up as needed.

>Adding a translator will add this translator as a visible option in the regular export dialog BTW. No way around that.

I think that wouldn't matter. Zutilo offers this function, but no translators are installed and indeed no extra quickcopies are enabled. I might then add those quickcopies and extra translators - but at that stage I've tweaked my system anyway through the advanced options. 

Would that be feasible?

**Comment by retorquere** [comment]
> Just to make sure I have understood. Atm, the Zotero translator is passed a structure as shown below.
> 
> ```
> {
>     "key": "W25GTPJU",
>     "version": 4774,
>     "library": {
>     },
>     "links": {
>     },
>     "meta": {
>     },
>     "data": {
>     }
> }
> ```

I don't recognize this structure. This is the result from `nextItem()`? I've never seen a `data` field on this. 

> Would it really trip up other translators if another field was inserted? I.e.

Nope, but if the suggestion was to have `nextItem()` also return items that *only* have `clickcontext` and either no `itemType` or an unexpected `itemType` (which is how I read the above, maybe mistaken), that would certainly break translators.

That's assuming you  can even get the clickcontext in there. I would currently not know how, and I'm pretty familiar with that path.

> Let's assume we only use 'quickcopy' and assume the 'translators is triggered through quickcopy' (but not by anything else).

That assumption doesn't hold. Translators will be available in the export dropdown. No way to prevent this.

> In other words: Zutilo offers a keyboard binding / right-click option for the standard quickcopy as well as two extra ones. (Right-click or keyboard shortcut.) If Zutilo is installed, then (only) if the right-click/keyboard shortcut is used, the additional 'clickcontext' is passed to the translator. E.g., the user may have the quickcopy set up for scannable cite. Would the extra data trip up scannable cite?

It's not extra data. It's the data that would be missing. In any case, what you describe here might be doable with postprocessing, which would solve both the breakage (if any) and the context-passing problem (because you're still in the click context at this point)
 
> > Adding a translator will add this translator as a visible option in the regular export dialog BTW. No way around that.
> 
> I think that wouldn't matter. Zutilo offers this function, but no translators are installed 

OK now I'm confused. We're talking about Zotero export translators? You can't use translators that aren't installed. I must be misunderstanding something.

**Comment by bjohas** [comment]
Ah yes - apologies, I exported that with zotero-cli. The actual data passed to the translator would be:
```
{
   "creators" : [
   ],
   "key" : "5KJK5NQQ",
   "itemID" : 46727,
...
   "attachments" : [
   ],
...
   "itemType" : "journalArticle",
   "language" : "en",
   "dateAdded" : "2020-02-19 20:52:33",
   "volume" : "35",
   "tags" : [
...
   "accessDate" : "2020-02-19 20:52:33"
// then add:
   "clickcontext" : {
      ...
  }
}
```

> OK now I'm confused. We're talking about Zotero export translators? You can't use translators that aren't installed. I must be misunderstanding something.

Yes, so this would be Zotero translators. But they would be meant to work only with Zutilo. The translators would handle the extra 'clickcontext' data. However, if the data 'clickcontext' data isn't there, then it doesn't matter. You just don't get (e.g.) the listing of the currently selected collection.

The missing data (when clicking on a collection): The click on the collectionl is only enabled through Zutilo. So I'd give you a translator that handles that click, but I'd say "This only works with Zutilo". The translator would be set up to be harmless when triggered on items, but give you the collection when used through Zutilo on a collection.

So yes, they are Zotero translators, but they offer special fuinction when called from Zutilo.

**Comment by retorquere** [comment]
> Ah yes - apologies, I exported that with zotero-cli. The actual data passed to the translator would be:

That would be fine. My "breakage" comment related to me interpreting the earlier proposal as injecting extra "pseudo-items" returned from `Zotero.nextItem()` inside translators.
 
> Yes, so this would be Zotero translators. But they would be meant to work only with Zutilo.

There's no way to enforce this. But if this is only meant to be used from a menu option, why use a translator? Generating a list of items in export format is trivial. I must still be overlooking something.

**Comment by bjohas** [comment]
> But if this is only meant to be used from a menu option, why use a translator? Generating a list of items in export format is trivial. I must still be overlooking something.

The translator makes that menu option configurable. So rather than asking for PR to Zutilo to implement 'my favourite export', the idea is to implement a more generic system, that people can tweak.

For example, I have a quick-copy-item-translator that produces the following text:
```
Author (year). TITLE. 
Zotero app:  zotero://select/2405685/items/RUSE8WYV; 
Browser: http://zotero.org/groups/2405685/items/RUSE8WYV; 
ETH library: https://docs.edtechhub.org/lib/RUSE8WYV; 
```
We use this in our research group when sharing papers. The rule is "If it's worth sharing with others, it goes into Zotero". You then use the translator to generate the text and paste it into an email or Slack or whatever. I can generate the above text just with a translator (from an item, so no problem).

Use case 1: However, I would now also add this to the right-click on collections (as per our extension to Zutilo).
```
Collection: 
Select: zotero://select/groups/2405685/collections/5CKG35JG
URI: http://zotero.org/groups/2405685/collections/5CKG35JG
ETH library: https://docs.edtechhub.org/lib/5CKG35JG
```
I cannot ask Zutilo to implement that... because it's very bespoke to us. Hence the idea of a translator-based system.

Use case 2: As per https://github.com/willsALMANJ/Zutilo/issues/91, we want to produce links that show the item in context, e.g. to produce this text:
```
Author (year). Title. DOI.
Selected collection: zotero://....
Additional collections:
- zotero://...
- zotero://...
- zotero://...
```
I think the personal preferences for how this is produced will vary a fair bit, and should be left to the authors of translators, rather than Zutilo.

Does that make sense?

**Comment by retorquere** [comment]
Somewhat, yeah. I can see why you'd want this, but that leaves unresolved how you're going to get that information into `itemToExportFormat`.

You can smuggle in extra data into an export translator by stuffing it into a displayOption, I use this for my auto export settings. That info can be just about anything, data-wise; if it would survive a json-stringify-parse, roughly speaking, it can be smuggled in that way. That's easy to do, easily picked up by a translator that knows about it, wouldn't break anything, and the customized translator could rebind the info so smuggled in based on the item key.

**Comment by bjohas** [comment]
> Somewhat, yeah. I can see why you'd want this, but that leaves unresolved how you're going to get that information into itemToExportFormat.

My assumption is that Zutilo can do this - from the right-click menu for the quickcopy-alt1/2 items. As those are implemented in Zutilo, I'm assuming that Zutilo would know the active collection. But maybe that's not the case?

**Comment by retorquere** [comment]
Zutilo would know, and can kick off a translator, but the items being serialized (itemToExportFormat) as part of that translation happens a few steps away from that point, and there's no robust way to bring the context you have at the kick off point to itemToExportFormat.

**Comment by retorquere** [comment]
This is in fact why saved-search auto-export isn't in BBT.

**Comment by bjohas** [comment]
Ahhh yes, you said that above - I now get it. OK, so it may just be hard to do via the translator route. Apologies for misunderstanding. Oooops :(

**The proposal** therefore (as intended originally) is then existing 'right-click' function in Zutilo (on collection/item) and extend this to 
- collect/item/search/library
- while also selecting items as appropriate.

I've made a google doc here:
https://docs.google.com/document/d/1ayD1ItSV8Sc3XWoayyO3eqzoygPXiKD-Zoohe7SO28o/edit?hl=en#

Basically, the functions to be enhanced are shown in the Zutilo prefs here:
![image](https://user-images.githubusercontent.com/7574634/75097277-be0ef100-55a0-11ea-9d43-297cdd33af87.png)

I'd like to propose to add a 3rd option called 'Copy select item and url' to each set of two. I often want to give people the choice which one they use, and tend to copy the item twice.

**Comment by retorquere** [comment]
> Ahhh yes, you said that above - I now get it. OK, so it may just be hard to do via the translator route. Apologies for misunderstanding. Oooops :(

No worries. It's still possible through the displayOptions. Not quite as elegent, but effective. displayOptions are set at the kickoff point, and are brought to the export translator by Zotero itself.

**Comment by bjohas** [comment]
Thank you @retorquere for the additional clarifications!

Basically, when the right-click quick-copy option is triggered, Zutilo would need to set, e.g.
```
setDisplayOptions({ clickcontext: { 
   type: "library|collections|searches", 
   key: "KEY", 
   group: "GROUPID", 
   items: ["AAA", "BBB"] } })
```
This could then be retrieved and used in the translator as
```
var context = Zotero.getHiddenOption(‘clickcontext');
var uri = "";
if (context) {
if (context.type != "library") {
uri = "zotero://select/groups/"
   + context.group + "/"
   + context.type + "/"
   + context.key 
   + "/items?itemKey="+context.items.join(",");
}
...
}
```
i.e. to produce
```
zotero://select/groups/GROUPID/searches/KEY/items?itemKey=AAA,BBB
```
(I guess the important parts are `context.type` and `context.key` as you cannot get those otherwise. `context.group` and `context.items` should be available via `nextItem()`.)

**Comment by retorquere** [comment]
for completeness: you'd [prepare a translator run](https://github.com/retorquere/zotero-better-bibtex/blob/master/content/translators.ts#L464), and then [call `translator.setDisplayOptions` on it](https://github.com/retorquere/zotero-better-bibtex/blob/master/content/translators.ts#L486) before [kicking that run off](https://github.com/retorquere/zotero-better-bibtex/blob/master/content/translators.ts#L523).
